packages/lesswrong/components/hooks/useExpandedFrontpageSection.tsx

```diff
@@ -0,0 +1,101 @@
+import { useState, useCallback } from "react";
+import { useCurrentUser } from "../common/withUser";
+import { useTracking } from "../../lib/analyticsEvents";
+import { useCookiesWithConsent } from "./useCookiesWithConsent";
+import { useMutation, gql } from "@apollo/client";
+import moment from "moment";
+
+export type ExpandedFrontpageSections = NonNullable<DbUser["expandedFrontpageSections"]>;
+export type ExpandedFrontpageSection = keyof ExpandedFrontpageSections;
+export type DefaultExpandedType = "all" | "none" | "loggedIn";
+
+export type UseExpandedFrontpageSectionProps = {
+  section: ExpandedFrontpageSection,
+  defaultExpanded: DefaultExpandedType,
+  onExpandEvent?: string,
+  onCollapseEvent?: string,
+  cookieName?: string,
+}
+
+const expandFrontpageSectionMutation = gql`
+  mutation UserExpandFrontpageSection($section: String!, $expanded: Boolean!) {
+    UserExpandFrontpageSection(section: $section, expanded: $expanded)
+  }
+`;
+
+const isDefaultExpanded = (
+  currentUser: UsersCurrent | null,
+  defaultExpanded: DefaultExpandedType,
+): boolean => {
+  switch (defaultExpanded) {
+  case "none":
+      return false;
+  case "all":
+      return true;
+  case "loggedIn":
+      return !!currentUser;
+  }
+}
+
+export const useExpandedFrontpageSection = ({
+  section,
+  defaultExpanded,
+  onExpandEvent,
+  onCollapseEvent,
+  cookieName,
+}: UseExpandedFrontpageSectionProps) => {
+  cookieName ??= `expand_frontpage_section_${section}`;
+
+  const currentUser = useCurrentUser();
+  const [expandFrontpageSection] = useMutation(expandFrontpageSectionMutation, {
+    errorPolicy: "all",
+    refetchQueries: ["getCurrentUser"],
+  });
+
+  const {captureEvent} = useTracking();
+  const [cookies, setCookie, removeCookie] = useCookiesWithConsent([cookieName]);
+
+  const userExpand = currentUser?.expandedFrontpageSections?.[section];
+  const cookieExpand = cookies[cookieName] && cookies[cookieName] !== "false";
+  const defaultExpand = isDefaultExpanded(currentUser, defaultExpanded);
+  const initialExpanded = userExpand ?? cookieExpand ?? defaultExpand ?? false;
+  const [expanded, setExpanded] = useState(initialExpanded);
```

Comment:
I noticed occasional issues here when I was testing while logged out (ex. sometimes the server and client would disagree, sometimes the arrow was not pointing the right way). But I couldn't consistently repro any of them, so maybe it's just my local. 🤷‍♀️ Not blocking, just something to keep an eye on.

Reply:
Hmm - I'm not able to repo, but I'll keep an eye on it. I think this kind of thing is easier to fix with a larger sample size 😅